query/resource_filter.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2022-2023 Searis AS
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package query
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// Filter describe a type can return an internal resource filter structure.
+// Filters reduce which resources.
+type Filter interface {
+	filter() filter
+}
+
+// And returns an new filter that merges the passed in filters with logical AND.
+func And(filters ...Filter) Filter {
+	newF := filter{
+		and: make([]filter, 0, len(filters)),
+	}
+	for _, ft := range filters {
+		f := ft.filter()
+		switch {
+		case len(f.or) == 0 && len(f.paths) == 0:
+			// Flatten and values (and skip empty queries).
+			newF.and = append(newF.and, f.and...)
+		default:
+			newF.and = append(newF.and, f)
+		}
+	}
+	if len(newF.and) == 1 {
+		return newF.and[0]
+	}
+	return newF
+}
+
+// Or returns an new filter that merges the passed in filters with logical OR.
+// Explicit nil values are omitted.
+func Or(filters ...Filter) Filter {
+	newF := filter{
+		or: make([]filter, 0, len(filters)),
+	}
+
+	for _, ft := range filters {
+
+		f := ft.filter()
+		if f.matchAll() {
+			// Optimization:
+			//   OR(matchAll,matchSome) == matchAll
+			return filter{}
+		}
+		switch {
+		case len(f.and) == 0 && len(f.paths) == 0:
+			// Flatten OR values (but only when the query is not empty).
+			newF.or = append(newF.or, f.or...)
+		default:
+			newF.or = append(newF.or, f)
+		}
+	}
+	if len(newF.or) == 1 {
+		return newF.or[0]
+	}
+	return newF
+}
+
+// filter describes a search filter for matching resources.
+type filter struct {
+	and   []filter
+	or    []filter
+	paths Comparisons
+}
+
+// All returns an empty filter, meaning it match all resources.
+func All() filter {
+	return filter{}
+}
+
+// matchAll return true if the filter matches all resources. A.k.a. the
+// filter is empty.
+func (f filter) matchAll() bool {
+	return len(f.and) == 0 && len(f.or) == 0 && len(f.paths) == 0
+}
+
+func (f filter) filter() filter { return f }
+
+var (
+	_ interface {
+		json.Marshaler
+		fmt.Stringer
+		Filter
+	} = filter{}
+	_ json.Unmarshaler = (*filter)(nil)
+)
```

Comment:
simplify here?

Reply:
This is a compile time test that we implement a specific set of interfaces. 

The following is equivalent:
```go
const (
	_ json.Marshaler = filter{}
	_ fmt.Stringer = filter{}
	_ Filter = filter{}
	_ json.Unmarshaler = (*filter)(nil)
)
```

I have landed on using the `_ interface{x,y,z} = T{}` pattern  consistently though.